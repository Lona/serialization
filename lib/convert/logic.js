const uuid = require('uuid/v4')

function createUUID() {
  return uuid().toUpperCase()
}

const singleChildMapping = {
  declaration: 'content',
  variable: 'initializer',
  literalExpression: 'literal',
  functionCallArgument: 'expression',
  memberExpression: 'expression',
}

const multipleChildMapping = {
  program: 'block',
  namespace: 'declarations',
}

const identifierNodeMapping = {
  importDeclaration: 'name',
  variable: 'name',
  namespace: 'name',
}

const annotationNodeMapping = {
  variable: 'annotation',
}

const upperFirst = string => string.slice(0, 1).toUpperCase() + string.slice(1)
const lowerFirst = string => string.slice(0, 1).toLowerCase() + string.slice(1)

function convertLogicJsonToXml(logicJson) {
  function getChildren(node) {
    // console.log(node)
    const { type, data } = node

    switch (type) {
      case 'functionCallExpression': {
        const { expression, arguments: args } = data
        console.log('ARGS', args)

        const mappedArgs = args.map(arg => arg.expression)

        return [expression, ...mappedArgs]
      }
      default:
        break
    }

    if (singleChildMapping[type]) {
      return [data[singleChildMapping[type]]]
    }
    if (multipleChildMapping[type]) {
      return data[multipleChildMapping[type]]
    }

    return []
  }

  function serializeAnnotationNode(node) {
    const { type, data } = node

    switch (type) {
      case 'typeIdentifier': {
        const { genericArguments, identifier } = data

        if (genericArguments && genericArguments.length > 0) {
          return `${identifier.string}[${genericArguments.join(',')}]`
        }

        return identifier.string
      }
      default:
        throw new Error(`Unhandled type identifier ${type}`)
    }
  }

  function processStandardNode(node) {
    const { type, data } = node

    const nodeName = upperFirst(type)

    const attributes = {}

    if (identifierNodeMapping[type]) {
      attributes.name = data[identifierNodeMapping[type]].name
    }

    if (annotationNodeMapping[type]) {
      attributes.type = serializeAnnotationNode(
        data[annotationNodeMapping[type]]
      )
    }

    switch (type) {
      case 'color':
      case 'number':
      case 'string':
      case 'boolean': {
        const { value } = data
        attributes.value = value
        break
      }
      case 'memberExpression': {
        const { memberName } = data
        attributes.name = memberName.string
        break
      }
      case 'identifierExpression': {
        const { identifier } = data
        attributes.name = identifier.string
        break
      }
      default:
        break
    }

    const children = getChildren(node).map(processStandardNode)

    return {
      name: nodeName,
      attributes,
      children,
    }
  }

  return logicJson.data.block.map(processStandardNode)
}

function convertLogicXmlToJson(program) {
  const { children: programStatements } = program

  function deserializeAnnotation(string) {
    return {
      type: 'typeIdentifier',
      data: {
        id: createUUID(),
        genericArguments: [],
        identifier: {
          id: createUUID(),
          isPlaceholder: false,
          string,
        },
      },
    }
  }

  function processStandardNode(node) {
    const { name, attributes = {}, children } = node

    const nodeName = lowerFirst(name)

    const data = {
      id: createUUID(),
      ...attributes,
    }

    if (singleChildMapping[nodeName]) {
      data[singleChildMapping[nodeName]] = processStandardNode(children[0])
    } else if (multipleChildMapping[nodeName]) {
      data[multipleChildMapping[nodeName]] = children.map(processStandardNode)
    }

    if (identifierNodeMapping[nodeName]) {
      data[identifierNodeMapping[nodeName]] = {
        id: createUUID(),
        name: attributes.name,
      }
    }

    if (annotationNodeMapping[nodeName]) {
      data[annotationNodeMapping[nodeName]] = deserializeAnnotation(
        attributes.type
      )
      delete data.type
    }

    switch (nodeName) {
      case 'number': {
        data.value = parseFloat(data.value)
        break
      }
      default:
        break
    }

    return {
      type: nodeName,
      data,
    }
  }

  return {
    type: 'program',
    data: {
      id: createUUID(),
      block: programStatements.map(processStandardNode),
    },
  }
}

module.exports = { convertLogicJsonToXml, convertLogicXmlToJson }
