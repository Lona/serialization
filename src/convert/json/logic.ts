import stringify from 'json-stable-stringify'
import Ajv from 'ajv'
import * as AST from '../../types/logic-ast'
import schema from './logic-schema.json'

export function print(ast: AST.SyntaxNode) {
  return stringify(ast, { space: '  ' })
}

export function parse(contents: string): AST.SyntaxNode {
  const ajv = new Ajv()
  const validate = ajv.compile(schema)
  const data = JSON.parse(contents)
  if (validate(data) && ajv.errors) {
    throw new Error(ajv.errors[0].message)
  }
  return data
}
