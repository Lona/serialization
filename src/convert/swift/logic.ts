import { v4 as uuid } from 'uuid'
import * as AST from '../../types/logic-ast'
import { indentBlock, repeat } from '../../formatting'
import { assertNever, rng } from '../../utils'
import parser from './pegjs/logicSwiftParser'

function noPlaceholder(x: AST.SyntaxNode) {
  return x.type !== 'placeholder'
}

export function parse(code: string, options?: {}): AST.SyntaxNode {
  return parser.parse(
    code,
    Object.assign({ generateId: () => uuid({ rng }) }, options)
  )
}

function printComment(node: AST.SyntaxNode, additionalComment: string = '') {
  if ('comment' in node.data && node.data.comment && node.data.comment.string) {
    return `/*
${node.data.comment.string
  .split('\n')
  .map(x => (x ? ` * ${x}` : ` *`))
  .join('\n')}${additionalComment ? `\n *\n${additionalComment}` : ''}
 */\n`
  }
  return ''
}

export function print(node: AST.SyntaxNode, options: { indent?: number } = {}) {
  const { indent = 2 } = options

  function printNode(node: AST.SyntaxNode): string {
    switch (node.type) {
      case 'program': {
        const { block } = node.data

        return block
          .filter(noPlaceholder)
          .map(printNode)
          .join('\n\n')
      }
      case 'topLevelDeclarations': {
        const { declarations } = node.data

        return declarations
          .filter(noPlaceholder)
          .map(printNode)
          .join('\n\n')
      }
      // Declaration statement
      case 'declaration': {
        return printNode(node.data.content)
      }
      case 'importDeclaration': {
        const { name } = node.data

        return `import ${name.name}`
      }
      case 'namespace': {
        const { name, declarations } = node.data

        const normalizedDeclarations = declarations
          .filter(noPlaceholder)
          .map(declaration => {
            return {
              ...declaration,
              data: { ...declaration.data, declarationModifier: 'static' },
            } as AST.Declaration
          })

        const printedDeclarations = normalizedDeclarations
          .map(x => printNode(x))
          .map(x => indentBlock(x, indent))
          .join('\n')

        return `${printComment(node)}extension ${name.name} {
${printedDeclarations}
}`
      }
      case 'record': {
        const { name, declarations } = node.data

        const printedDeclarations = declarations
          .filter(noPlaceholder)
          .map(x => printNode(x))
          .map(x => indentBlock(x, indent))
          .join('\n')

        const generics = node.data.genericParameters.filter(noPlaceholder)

        return `${printComment(node)}struct ${name.name}${
          generics.length ? `<${generics.map(printNode).join(', ')}>` : ''
        } {
${printedDeclarations}
}`
      }
      case 'variable': {
        const { annotation, initializer, name, declarationModifier } = node.data

        const printedDeclarationModifier = declarationModifier
          ? `${declarationModifier} `
          : ''
        const printedAnnotation = annotation ? `: ${printNode(annotation)}` : ''
        const printedInitializer = initializer
          ? ` = ${printNode(initializer)}`
          : ''

        return `${printComment(node)}${printedDeclarationModifier}let ${
          name.name
        }${printedAnnotation}${printedInitializer}`
      }
      case 'typeIdentifier': {
        const { genericArguments, identifier } = node.data

        return genericArguments.filter(noPlaceholder).length > 0
          ? `${identifier.string}<${genericArguments
              .filter(noPlaceholder)
              .map(printNode)
              .join(', ')}>`
          : identifier.string
      }
      case 'functionCallExpression': {
        const { expression } = node.data

        const printedArguments = node.data.arguments
          .filter(noPlaceholder)
          .map(printNode)
          .join(', ')

        return `${printNode(expression)}(${printedArguments})`
      }
      case 'argument': {
        const { expression, label } = node.data

        return label
          ? `${label}: ${printNode(expression)}`
          : printNode(expression)
      }
      case 'memberExpression': {
        const { expression, memberName } = node.data

        return printNode(expression) + '.' + memberName.string
      }
      case 'identifierExpression': {
        const { identifier } = node.data

        return identifier.string
      }
      case 'literalExpression': {
        const { literal } = node.data

        return printNode(literal)
      }
      case 'boolean':
      case 'number': {
        const { value } = node.data

        return value.toString()
      }
      case 'string': {
        const { value } = node.data

        return JSON.stringify(value)
      }
      case 'color': {
        const { value } = node.data

        return `#color(css: ${JSON.stringify(value)})`
      }
      case 'array': {
        const { value } = node.data

        const printedExpressions = value
          .filter(noPlaceholder)
          .map(printNode)
          .map(x => indentBlock(x, indent))
          .join(',\n')

        return `[
${printedExpressions}
]`
      }
      case 'loop': {
        return `while ${printNode(node.data.expression)} {
${node.data.block
  .filter(noPlaceholder)
  .map(printNode)
  .map(x => indentBlock(x, indent))
  .join('\n')}
}`
      }
      case 'branch': {
        return `if ${printNode(node.data.condition)} {
${node.data.block
  .filter(noPlaceholder)
  .map(printNode)
  .map(x => indentBlock(x, indent))
  .join('\n')}
}`
      }
      case 'expression': {
        return printNode(node.data.expression)
      }
      case 'function': {
        const generics = node.data.genericParameters.filter(noPlaceholder)
        const params = node.data.parameters.filter(noPlaceholder)

        const paramsComment = params
          .map(x => {
            if (x.type !== 'parameter') {
              return ''
            }

            if (!x.data.comment || !x.data.comment.string) {
              return ''
            }
            let comment = ` * @param ${x.data.localName.name}`
            const indent = `@param ${x.data.localName.name} - `.length
            comment += ` - ${x.data.comment.string
              .split('\n')
              .map((l, i) =>
                i === 0 ? l : l ? ` * ${repeat(' ', indent)}${l}` : ` *`
              )
              .join('\n')}`

            return comment
          })
          .filter(x => !!x)
          .join('\n')

        const body = node.data.block.filter(noPlaceholder)

        return `${printComment(node, paramsComment)}func ${
          node.data.name.name
        }${
          generics.length ? `<${generics.map(printNode).join(', ')}>` : ''
        }(${params.map(printNode).join(', ')}) -> ${printNode(
          node.data.returnType
        )} ${
          body.length
            ? `{
${node.data.block
  .filter(noPlaceholder)
  .map(printNode)
  .map(x => indentBlock(x, indent))
  .join('\n')}
}`
            : `{}`
        }`
      }
      case 'enumeration': {
        const generics = node.data.genericParameters.filter(noPlaceholder)
        return `${printComment(node)}enum ${node.data.name.name}${
          generics.length ? `<${generics.map(printNode).join(', ')}>` : ''
        } {
${node.data.cases
  .filter(noPlaceholder)
  .map(printNode)
  .map(x => indentBlock(x, indent))
  .join('\n')}
}`
      }
      case 'enumerationCase': {
        const valueTypes = node.data.associatedValueTypes.filter(noPlaceholder)
        return `${printComment(node)}case ${
          node.data.name.name
        }(${valueTypes.map(printNode).join(', ')})`
      }
      case 'assignmentExpression': {
        return `${printNode(node.data.left)} = ${printNode(node.data.right)}`
      }
      case 'placeholder':
        return ''
      case 'parameter': {
        if ('annotation' in node.data) {
          return `${node.data.localName.name}: ${printNode(
            node.data.annotation
          )}${
            node.data.defaultValue.type === 'value'
              ? ` = ${printNode(node.data.defaultValue)}`
              : ''
          }`
        }
        return node.data.name.name
      }
      case 'none': {
        return ''
      }
      case 'value': {
        return printNode(node.data.expression)
      }
      case 'functionType': {
        return `(${node.data.argumentTypes
          .filter(noPlaceholder)
          .map(printNode)
          .join(', ')}) -> ${printNode(node.data.returnType)}`
      }
      case 'topLevelParameters': {
        // TODO: what is that?
        return ''
      }
      case 'return': {
        return `return ${printNode(node.data.expression)}`
      }
      default:
        console.log(node)
        assertNever(node)
    }
    return ''
  }

  return printNode(node)
}
