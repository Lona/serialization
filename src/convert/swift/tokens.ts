import { indentBlock } from '../../formatting'
import parser from './pegjs/logicSwiftParser'

export function parse(code: string, options?: {}) {
  return parser.parse(code, options)
}

export function print(node, options: { indent?: number } = {}) {
  const { indent = 2 } = options

  function printNode(node) {
    const { data } = node

    switch (node.type) {
      case 'program': {
        const { block } = data

        return block
          .filter(node => node.type !== 'placeholder')
          .map(printNode)
          .join('\n\n')
      }
      case 'topLevelDeclarations': {
        const { declarations } = data

        return declarations
          .filter(node => node.type !== 'placeholder')
          .map(printNode)
          .join('\n\n')
      }
      // Declaration statement
      case 'declaration': {
        const { content } = data

        return printNode(content)
      }
      case 'importDeclaration': {
        const {
          name: { name },
        } = data

        return `import ${name}`
      }
      case 'namespace': {
        const {
          name: { name },
          declarations,
        } = data

        const normalizedDeclarations = declarations.map(declaration => {
          return {
            ...declaration,
            data: { ...declaration.data, declarationModifier: 'static' },
          }
        })

        const printedDeclarations = normalizedDeclarations
          .filter(declaration => declaration.type !== 'placeholder')
          .map(x => printNode(x))
          .map(x => indentBlock(x, indent))
          .join('\n')

        return `enum ${name} {
${printedDeclarations}
}`
      }
      case 'record': {
        const {
          name: { name },
          declarations,
        } = data

        const printedDeclarations = declarations
          .filter(declaration => declaration.type !== 'placeholder')
          .map(x => printNode(x))
          .map(x => indentBlock(x, indent))
          .join('\n')

        return `struct ${name} {
${printedDeclarations}
}`
      }
      case 'variable': {
        const {
          annotation,
          initializer,
          name: { name },
          declarationModifier,
        } = data

        const printedDeclarationModifier = declarationModifier
          ? declarationModifier + ' '
          : ''

        return `${printedDeclarationModifier}let ${name}: ${printNode(
          annotation
        )} = ${printNode(initializer)}`
      }
      case 'typeIdentifier': {
        const {
          genericArguments,
          identifier: { string },
        } = data

        return genericArguments.length > 0
          ? `${string}<${genericArguments.map(printNode).join(', ')}>`
          : string
      }
      case 'functionCallExpression': {
        const { expression } = data

        const printedArguments = data.arguments
          .filter(node => node.type !== 'placeholder')
          .map(printNode)
          .join(', ')

        return `${printNode(expression)}(${printedArguments})`
      }
      case 'argument': {
        const { expression, label } = data

        return label
          ? `${label}: ${printNode(expression)}`
          : printNode(expression)
      }
      case 'memberExpression': {
        const {
          expression,
          memberName: { string },
        } = data

        return printNode(expression) + '.' + string
      }
      case 'identifierExpression': {
        const {
          identifier: { string },
        } = data

        return string
      }
      case 'literalExpression': {
        const { literal } = data

        return printNode(literal)
      }
      case 'boolean':
      case 'number': {
        const { value } = data

        return value.toString()
      }
      case 'string': {
        const { value } = data

        return JSON.stringify(value)
      }
      case 'color': {
        const { value } = data

        return `#color(css: ${JSON.stringify(value)})`
      }
      case 'array': {
        const { value } = data

        const printedExpressions = value
          .filter(node => node.type !== 'placeholder')
          .map(printNode)
          .map(x => indentBlock(x, indent))
          .join('\n\n')

        return `[
${printedExpressions}
]`
      }
    }
  }

  return printNode(node)
}
