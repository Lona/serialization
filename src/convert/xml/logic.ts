import uuid from 'uuid/v4'
import { parseXML, buildXML, XMLNode } from './utils'
import * as AST from '../../types/logic-ast'

function createUUID() {
  return uuid().toUpperCase()
}

enum literalToTypeMapping {
  boolean = 'Boolean',
  number = 'Number',
  string = 'String',
  color = 'Color',
}
const typeToLiteralMapping = Object.entries(literalToTypeMapping).reduce(
  (result, [key, value]) => {
    result[value] = key
    return result
  },
  {} as { [key: string]: string }
)

const singleChildMapping: { [key: string]: string } = {
  declaration: 'content',
  variable: 'initializer',
  literalExpression: 'literal',
  argument: 'expression',
  memberExpression: 'expression',
}

const multipleChildMapping: { [key: string]: string } = {
  program: 'block',
  namespace: 'declarations',
  topLevelDeclarations: 'declarations',
  record: 'declarations',
  functionCallExpression: 'arguments',
}

const implicitPlaceholderMapping: { [key: string]: string } = {
  program: 'block',
  namespace: 'declarations',
  topLevelDeclarations: 'declarations',
  record: 'declarations',
  functionCallExpression: 'arguments',
}

const nodeRenaming: { [key: string]: string } = {
  topLevelDeclarations: 'Declarations',
}

const reverseNodeRenaming: { [key: string]: string } = {
  Declarations: 'topLevelDeclarations',
}

const patternNodeMapping: { [key: string]: string } = {
  importDeclaration: 'name',
  variable: 'name',
  namespace: 'name',
}

const identifierNodeMapping: { [key: string]: string } = {
  memberExpression: 'memberName',
}

const annotationNodeMapping: { [key: string]: string } = {
  variable: 'annotation',
}

const upperFirst = (string: string) =>
  string.slice(0, 1).toUpperCase() + string.slice(1)
const lowerFirst = (string: string) =>
  string.slice(0, 1).toLowerCase() + string.slice(1)

export function print(logicJson: AST.SyntaxNode): string {
  function isNotPlaceholder(
    item: AST.SyntaxNode,
    index: number,
    list: AST.SyntaxNode[]
  ) {
    if (
      index === list.length - 1 &&
      'type' in item &&
      item.type === 'placeholder'
    ) {
      return false
    }
    return true
  }

  function serializeAnnotationNode(node: AST.TypeAnnotation): string {
    switch (node.type) {
      case 'typeIdentifier': {
        const { genericArguments, identifier } = node.data

        if (genericArguments && genericArguments.length > 0) {
          const serializedArguments = genericArguments.map(
            serializeAnnotationNode
          )
          return `${identifier.string}(${serializedArguments.join(',')})`
        }

        return identifier.string
      }
      default:
        throw new Error(`Unhandled type identifier ${node.type}`)
    }
  }

  function processStandardNode(node: AST.SyntaxNode): XMLNode {
    const nodeName = nodeRenaming[node.type] || upperFirst(node.type)

    const attributes: {
      name?: string
      type?: string
      label?: string
      value?: any
    } = {}

    if ('data' in node && patternNodeMapping[node.type]) {
      // @ts-ignore
      attributes.name = node.data[patternNodeMapping[node.type]].name
    }

    if ('data' in node && identifierNodeMapping[node.type]) {
      // @ts-ignore
      attributes.name = node.data[identifierNodeMapping[node.type]].string
    }

    if ('data' in node && annotationNodeMapping[node.type]) {
      attributes.type = serializeAnnotationNode(
        // @ts-ignore
        node.data[annotationNodeMapping[node.type]]
      )
    }

    switch (node.type) {
      case 'argument': {
        const { label } = node.data
        attributes.label = label
        break
      }
      case 'variable': {
        const compactLiteralTypes = ['boolean', 'number', 'string', 'color']

        if (
          node.data.initializer &&
          node.data.initializer.type === 'literalExpression'
        ) {
          const literal = node.data.initializer.data.literal
          if (
            compactLiteralTypes.includes(literal.type) &&
            'value' in literal.data
          ) {
            return {
              name: nodeName,
              attributes: {
                ...attributes,
                value: literal.data.value,
              },
              children: [],
            }
          }
          if (literal.type === 'array') {
            return {
              name: nodeName,
              attributes: {
                ...attributes,
              },
              children: literal.data.value
                .filter(isNotPlaceholder)
                .map(processStandardNode),
            }
          }
        }

        break
      }
      case 'record': {
        const {
          name: { name },
        } = node.data

        attributes.name = name
        break
      }
      case 'declaration': {
        const child = processStandardNode(node.data.content)

        return {
          ...child,
          name: ['Declaration', child.name].join('.'),
        }
      }
      case 'color':
      case 'number':
      case 'string':
      case 'boolean': {
        const { value } = node.data
        attributes.value = value
        break
      }
      case 'literalExpression': {
        const literalNode = processStandardNode(node.data.literal)

        return {
          name: 'Literal',
          attributes: {
            type: literalNode.name,
            value: literalNode.attributes.value,
          },
        }
      }
      case 'identifierExpression': {
        const { identifier } = node.data
        attributes.name = identifier.string
        break
      }
      default:
        break
    }

    const children = AST.subNodes(node).map(processStandardNode)

    return {
      name: nodeName,
      attributes,
      children,
    }
  }

  return buildXML(processStandardNode(logicJson))
}

export function parse(root: string): AST.SyntaxNode {
  function makePlaceholder() {
    return {
      data: { id: createUUID() },
      type: 'placeholder',
    } as { type: 'placeholder'; data: { id: string } }
  }

  const compactLiteralTypes = ['Boolean', 'Number', 'String', 'Color']

  function decodeLiteralValue(type: string, value: any) {
    switch (type) {
      case 'Boolean':
        return value === 'true'
      case 'Number':
        return value
      case 'String':
        return value
      case 'Color':
        return value
      default:
        throw new Error('Invalid literal value type')
    }
  }

  // TODO: Handle nested generics
  function deserializeAnnotation(
    string: string
  ): AST.TypeIdentifierTypeAnnotation {
    const startParens = string.indexOf('(')
    const endParens = string.lastIndexOf(')')

    let genericArguments = [] as AST.TypeAnnotation[]

    const name = startParens >= 0 ? string.slice(0, startParens) : string
    if (startParens !== -1 && endParens !== -1) {
      const argumentsString = string.slice(startParens + 1, endParens)
      const args = argumentsString.split(',').map(s => s.trim())
      genericArguments = args.map(deserializeAnnotation)
    }

    return {
      type: 'typeIdentifier',
      data: {
        id: createUUID(),
        genericArguments,
        identifier: {
          id: createUUID(),
          isPlaceholder: false,
          string: name,
        },
      },
    }
  }

  function processStandardNode(node: XMLNode): AST.SyntaxNode {
    const { name, attributes = {}, children = [] } = node

    switch (name) {
      case 'IdentifierExpression':
        return {
          data: {
            id: createUUID(),
            identifier: {
              id: createUUID(),
              isPlaceholder: false,
              string: attributes.name,
            },
          },
          type: 'identifierExpression',
        }
      case 'Record':
        return {
          data: {
            declarations: [
              ...children.map(processStandardNode).filter(AST.isDeclaration),
              {
                data: { id: createUUID() },
                type: 'placeholder',
              },
            ],
            genericParameters: [],
            id: createUUID(),
            name: {
              id: createUUID(),
              name: attributes.name,
            },
          },
          type: 'record',
        }
      case 'Declaration.ImportDeclaration':
        return {
          type: 'declaration',
          data: {
            id: createUUID(),
            content: processStandardNode({
              ...node,
              name: 'ImportDeclaration',
            }) as AST.ImportDeclaration,
          },
        }
      case 'Declaration.Namespace':
        return {
          type: 'declaration',
          data: {
            id: createUUID(),
            content: processStandardNode({
              ...node,
              name: 'Namespace',
            }) as AST.NamespaceDeclaration,
          },
        }
      case 'Literal':
        return processStandardNode({
          name: 'LiteralExpression',
          attributes: {},
          children: [
            {
              name: attributes.type,
              attributes: {
                value: attributes.value,
              },
              children: [],
            },
          ],
        })
      case 'Variable': {
        if (compactLiteralTypes.includes(attributes.type) && attributes.value) {
          return processStandardNode({
            name: 'Variable',
            attributes: {
              name: attributes.name,
              type: attributes.type,
            },
            children: [
              {
                name: 'LiteralExpression',
                attributes: {},
                children: [
                  {
                    name: typeToLiteralMapping[attributes.type],
                    attributes: {
                      value: decodeLiteralValue(
                        attributes.type,
                        attributes.value
                      ),
                    },
                    children: [],
                  },
                ],
              },
            ],
          })
        }

        const deserializedType = deserializeAnnotation(attributes.type)

        if (deserializedType.data.identifier.string === 'Array') {
          return {
            type: 'variable',
            data: {
              id: createUUID(),
              name: {
                id: createUUID(),
                name: attributes.name,
              },
              annotation: deserializedType,
              initializer: {
                type: 'literalExpression',
                data: {
                  id: createUUID(),
                  literal: {
                    type: 'array',
                    data: {
                      id: createUUID(),
                      value: [
                        ...children
                          .map(processStandardNode)
                          .filter(AST.isExpression),
                        makePlaceholder(),
                      ],
                    },
                  },
                },
              },
            },
          }
        }

        break
      }
      case 'Declaration.Variable': {
        if (compactLiteralTypes.includes(attributes.type) && attributes.value) {
          return {
            type: 'declaration',
            data: {
              id: createUUID(),
              content: processStandardNode({
                name: 'Variable',
                attributes: {
                  name: attributes.name,
                  type: attributes.type,
                },
                children: [
                  {
                    name: 'LiteralExpression',
                    attributes: {},
                    children: [
                      {
                        name: typeToLiteralMapping[attributes.type],
                        attributes: {
                          value: decodeLiteralValue(
                            attributes.type,
                            attributes.value
                          ),
                        },
                        children: [],
                      },
                    ],
                  },
                ],
              }) as AST.VariableDeclaration,
            },
          }
        }

        return {
          type: 'declaration',
          data: {
            id: createUUID(),
            content: processStandardNode({
              ...node,
              name: 'Variable',
            }) as AST.VariableDeclaration,
          },
        }
      }
      default:
        break
    }

    const nodeName = reverseNodeRenaming[name] || lowerFirst(name)

    // We implicitly transfer any single-value nodes to the data object
    const data: { [key: string]: any } = {
      id: createUUID(),
      ...attributes,
    }

    if (nodeName === 'functionCallExpression') {
      const [expression, ...args] = children
      data.expression = processStandardNode(expression)
      const processedArguments = args.map(processStandardNode)
      data.arguments =
        processedArguments.length === 0
          ? processedArguments
          : [...processedArguments, makePlaceholder()]
    } else if (singleChildMapping[nodeName]) {
      data[singleChildMapping[nodeName]] = processStandardNode(children[0])
    } else if (multipleChildMapping[nodeName]) {
      data[multipleChildMapping[nodeName]] = children.map(processStandardNode)

      if (implicitPlaceholderMapping[nodeName]) {
        data[multipleChildMapping[nodeName]].push(makePlaceholder())
      }
    }

    if (patternNodeMapping[nodeName]) {
      data[patternNodeMapping[nodeName]] = {
        id: createUUID(),
        name: attributes.name,
      }
    }

    if (identifierNodeMapping[nodeName]) {
      delete data.name
      data[identifierNodeMapping[nodeName]] = {
        id: createUUID(),
        isPlaceholder: false,
        string: attributes.name,
      }
    }

    if (annotationNodeMapping[nodeName]) {
      data[annotationNodeMapping[nodeName]] = deserializeAnnotation(
        attributes.type
      )
      delete data.type
    }

    switch (nodeName) {
      case 'number': {
        data.value = parseFloat(data.value)
        break
      }
      default:
        break
    }

    return {
      // @ts-ignore
      type: nodeName,
      // @ts-ignore
      data,
    }
  }

  return processStandardNode(parseXML(root))
}
