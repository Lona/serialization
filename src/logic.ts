import * as AST from './ast/logicAst'

import * as json from './convert/json/logic'
import * as swift from './convert/swift/logic'

import { normalizeFormat, SERIALIZATION_FORMAT } from './format'
import { parsingError } from './utils'

export function decodeLogic(
  contents: string,
  format?: SERIALIZATION_FORMAT,
  options: { filePath?: string; offset?: number } = {}
) {
  const sourceFormat = normalizeFormat(contents, format)
  try {
    switch (sourceFormat) {
      case SERIALIZATION_FORMAT.JSON:
        return json.parse(contents)
      case SERIALIZATION_FORMAT.SOURCE:
        return swift.parse(contents)
      default:
        throw new Error(`Unknown decoding format ${sourceFormat}`)
    }
  } catch (e) {
    if (e.name === 'SyntaxError' && e.location) {
      // that's a pegjs error, let's format it nicely
      throw parsingError(contents, e, options)
    }
    throw new Error(`Failed to decode logic as ${sourceFormat}.\n\n${e}`)
  }
}

export function encodeLogic(ast: AST.SyntaxNode, format: SERIALIZATION_FORMAT) {
  try {
    switch (format) {
      case SERIALIZATION_FORMAT.JSON:
        return json.print(ast)
      case SERIALIZATION_FORMAT.SOURCE:
        return swift.print(ast)
      default:
        throw new Error(`Unknown encoding format ${format}`)
    }
  } catch (e) {
    console.error(e)
    throw new Error(`Failed to encode logic as ${format}.\n\n${e}`)
  }
}

export function convertLogic(
  contents: string,
  targetFormat: SERIALIZATION_FORMAT,
  options: {
    sourceFormat?: SERIALIZATION_FORMAT
    filePath?: string
    offset?: number
  } = {}
) {
  const sourceFormat = normalizeFormat(contents, options.sourceFormat)

  if (sourceFormat === targetFormat) return contents

  const ast = decodeLogic(contents, sourceFormat, options)
  return encodeLogic(ast, targetFormat)
}
