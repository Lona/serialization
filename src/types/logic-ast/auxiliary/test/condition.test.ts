import { parse, print } from '../../../../convert/swift/logic'

jest.mock('uuid', () => ({ v4: () => '0' }))

describe('condition', () => {
  const parserOptions = { startRule: 'condition' }

  it('expression condition', () => {
    const code = `a`
    const ast = parse(code, parserOptions)
    expect(ast).toMatchSnapshot()
    expect(print(ast)).toBe(code)
  })

  it('case condition', () => {
    const code = `case Test.value(x) = y`
    const ast = parse(code, parserOptions)
    expect(ast).toMatchSnapshot()
    expect(print(ast)).toBe(code)
  })

  describe('invalid case condition', () => {
    it('missing parentheses', () => {
      const code = `case Test.value = y`
      expect(() => parse(code, parserOptions)).toThrow()
    })

    it('missing type identifier', () => {
      const code = `case .value = y`
      expect(() => parse(code, parserOptions)).toThrow()
    })

    it('missing initializer', () => {
      const code = `case Test.value`
      expect(() => parse(code, parserOptions)).toThrow()
    })
  })
})
