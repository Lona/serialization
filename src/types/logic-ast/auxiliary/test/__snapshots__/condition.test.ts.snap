// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`condition case condition 1`] = `
Object {
  "data": Object {
    "id": "0",
    "initializer": Object {
      "data": Object {
        "id": "0",
        "identifier": Object {
          "id": "0",
          "isPlaceholder": false,
          "string": "y",
        },
      },
      "type": "identifierExpression",
    },
    "pattern": Object {
      "caseName": Object {
        "id": "0",
        "isPlaceholder": false,
        "string": "value",
      },
      "id": "0",
      "tuple": Object {
        "elements": Array [
          Object {
            "data": Object {
              "id": "0",
              "name": "x",
            },
            "type": "identifierPattern",
          },
        ],
        "id": "0",
      },
      "typeIdentifier": Object {
        "data": Object {
          "genericArguments": Array [],
          "id": "0",
          "identifier": Object {
            "id": "0",
            "isPlaceholder": false,
            "string": "Test",
          },
        },
        "type": "typeIdentifier",
      },
    },
  },
  "type": "caseCondition",
}
`;

exports[`condition expression condition 1`] = `
Object {
  "data": Object {
    "expression": Object {
      "data": Object {
        "id": "0",
        "identifier": Object {
          "id": "0",
          "isPlaceholder": false,
          "string": "a",
        },
      },
      "type": "identifierExpression",
    },
    "id": "0",
  },
  "type": "expressionCondition",
}
`;
